cherab/tools/observers/bolometry.py

```diff
@@ -276,229 +227,151 @@ def __init__(self, detector_id, centre_point, basis_x, dx, basis_y, dy, slit,
-
-            self._volume_observer.observe()
-            self._volume_radiance_sensitivity.sensitivity[i] = self._volume_radiance_pipeline.value.mean
-            self._volume_power_sensitivity.sensitivity[i] = self._volume_power_pipeline.value.mean
-
-            outer_cylinder.parent = None
-            inner_cylinder.parent = None
-            cell_emitter.parent = None
-
-    def calculate_etendue(self, ray_count=10000, batches=10, print_results=False):
-
-        if self.slit.csg_aperture is None:
-            raise ValueError("CSG aperture is required to support etendue calculation.")
+    def trace_sightline(self):
+
+        if not isinstance(self.root, World):
+            raise ValueError("This BolometerFoil is not connected to a valid World scenegraph object.")
+
+        centre_point = self.centre_point
+
+        while True:
+
+            # Find the next intersection point of the ray with the world
+            intersection = self.root.hit(CoreRay(centre_point, self.sightline_vector))
+
+            if intersection is None:
+                raise RuntimeError("No material intersection was found for this sightline.")
+
+            elif isinstance(intersection.primitive.material, NullMaterial):
+                centre_point += self.sightline_vector * 1E-4
+                continue
+
+            else:
+                hit_point = intersection.hit_point.transform(intersection.primitive_to_world)
+                return self.centre_point, hit_point, intersection.primitive
+
+    def calculate_sensitivity(self, voxel_collection, ray_count=10000):
+
+        if not isinstance(voxel_collection, VoxelCollection):
+            raise TypeError("voxel_collection must be of type VoxelCollection")
+
+        if self.units == "Power":
+            pipeline = SpectralPowerPipeline0D(display_progress=False)
+        elif self.units == "Radiance":
+            pipeline = SpectralRadiancePipeline0D(display_progress=False)
+        else:
+            raise ValueError("Sensitivity units can only be of type 'Power' or 'Radiance'.")
+
+        voxel_collection.set_active("all")
+
+        cached_max_wavelength = self.max_wavelength
+        cached_min_wavelength = self.min_wavelength
+        cached_bins = self.spectral_bins
+        cached_pipelines = self.pipelines
+        cached_ray_count = self.pixel_samples
+
+        self.pipelines = [pipeline]
+        self.min_wavelength = 1
+        self.max_wavelength = voxel_collection.count + 1
+        self.spectral_bins = voxel_collection.count
+        self.pixel_samples = ray_count
+
+        self.observe()
+
+        self.max_wavelength = cached_max_wavelength
+        self.min_wavelength = cached_min_wavelength
+        self.spectral_bins = cached_bins
+        self.pipelines = cached_pipelines
+        self.pixel_samples = cached_ray_count
+
+        return pipeline.samples.mean
+
+    def calculate_etendue(self, ray_count=10000, batches=10):
 
         if not batches > 5:
```

Comment:
Why not `batches < 5` instead of `not batches > 5`? Pylint always moans at me for this. And in any case, your error message is slightly wrong. This condition as it stands is enforcing a minimum batch size of 6, not 5.

Reply:
agreed, made the change